lib/accessmonitoring/monitor.go

```diff
@@ -0,0 +1,287 @@
+func NewAccessMonitor(cfg Config) (*AccessMonitor, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &AccessMonitor{
+		cfg: cfg,
+	}, nil
+}
+
+// AddAccessMonitoringRuleHandler appends the access monitoring rule event handler.
+// These handlers are executed whenever a new access monitoring rule event is
+// observed.
+func (s *AccessMonitor) AddAccessMonitoringRuleHandler(handler EventHandler) {
+	s.ruleHandlers = append(s.ruleHandlers, handler)
+}
+
+// AddAccessRequestHandler appends the access request handler.
+// These handlers are executed whenever a new access request event is observed.
+func (s *AccessMonitor) AddAccessRequestHandler(handler EventHandler) {
+	s.requestHandlers = append(s.requestHandlers, handler)
+}
+
+// Run runs the access monitor. Restarts on failure.
+func (s *AccessMonitor) Run(ctx context.Context) error {
+	const retryJitter = 10 * time.Second
+	const retryInterval = 30 * time.Second
+	const lockTTL = time.Minute
+
+	waitWithJitter := retryutils.SeventhJitter(retryJitter)
+	for {
+		err := backend.RunWhileLocked(ctx, backend.RunWhileLockedConfig{
+			LockConfiguration: backend.LockConfiguration{
+				Backend:            s.cfg.Backend,
+				LockNameComponents: []string{componentName},
+				TTL:                lockTTL,
+				RetryInterval:      retryInterval,
+			},
+		}, s.run)
+		if err != nil {
+			s.cfg.Logger.ErrorContext(
+				ctx,
+				"Encountered a fatal error, it will restart after backoff.",
+				"component", componentName,
+				"error", err,
+				"restart_after", waitWithJitter,
+			)
+			if s.cfg.FailFast {
+				return trace.Wrap(err)
+			}
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(waitWithJitter):
+		}
+	}
+}
+
+func (s *AccessMonitor) run(ctx context.Context) error {
+	// AllowPartialSuccess is set to true because previous versions of access
+	// monitoring plugins did not require permissions to read access monitoring rule.
+	// TODO(bernardjkim): Migrate users onto access monitoring rules and disallow
+	// partial success.
+	const allowPartialSuccessTrue = true
+
+	watchKinds := []types.WatchKind{
+		{Kind: types.KindAccessRequest},
+		{Kind: types.KindAccessMonitoringRule},
+	}
+
+	// Initialize the watcher.
+	watcher, err := s.cfg.Events.NewWatcher(ctx, types.Watch{
+		Name:                componentName,
+		Kinds:               watchKinds,
+		AllowPartialSuccess: allowPartialSuccessTrue,
+	})
+	if err != nil {
+		return trace.Wrap(err, "failed to create watcher")
+	}
+	defer func() {
+		if err := watcher.Close(); err != nil {
+			s.cfg.Logger.ErrorContext(ctx, "Failed to close watcher.", "error", err)
+		}
+	}()
+
+	// Watcher is open, expect init event.
+	select {
+	case initEvent := <-watcher.Events():
+		if initEvent.Type != types.OpInit {
+			return trace.BadParameter("watcher yielded %[1]v (%[1]d) as first event, expected Init (this is a bug)", initEvent.Type)
```

Comment:
Is this string valid for string formatting?
It has two placeholders (`%[1]v` and `%[1]d`), but only one value (`initEvent.Type`)

Reply:
I believe so. The `%[1]` should resolve to the argument at index 1. Just to double check, this would return something like: `watcher yielded Delete (2) as first event, expected Init (this is a bug)`.